Add versionstamp documentation (#2554)
apple/foundationdb: documentation/sphinx/source/data-modeling.rst

            cols[c] = v
        return cols

Versionstamps
-------------

> A common data model is to index your data with a sequencing prefix to allow log scans or tails of recent data. This index requires a unique, monotonically increasing value. This could be implemented at the client level by reading the value for conflict checks before every increment. A better solution is the versionstamp, which can be generated at commit-time with no read conflict ranges, providing a unique sequence ID in a single conflict-free write.

Reviewer: I think it'd be worth making your allusion more explicit that versionstamps are the replacement for AUTOINCREMENT PRIMARY KEY.

Author: Got it.  versionstamp is FDB's version of to an AUTOINCREMENT PRIMARY KEY in SQL

Reviewer 2: Is there a change that is associated with this comment?

Author: Whoops, there was in my head!  Just pushed

> The version is used by FoundationDB to provide MVCC guarantees and transactional integrity. Versionstamps write the transaction's commit version as a value to an arbitrary key as part of the same transaction, allowing the client to leverage the version's unique and serial properties.  Because the versionstamp is generated at commit-time, the versionstamped key cannot be read in the same transaction that it is written, and the versionstamp's value will be unknown until the transaction is committed. After the transaction is committed, the versionstamp can be obtained.

Reviewer: ```suggestion
Versioning commits provides FoundationDB with MVCC guarantees and transactional integrity. Versionstamps write the transaction's commit version as a value to an arbitrary key as part of the same transaction, allowing the client to leverage the version's unique and serial properties.  Because the versionstamp is generated at commit-time, the versionstamped key cannot be read in the same transaction that it is written, and the versionstamp's value will be unknown until the transaction is committed. After the transaction is committed, the versionstamp can be obtained.
```

The versionstamp guarantees uniqueness and monotonically increasing values for the entire lifetime of a single FDB cluster. This is even true if the cluster is restored from a backup, as a restored cluster will begin at a higher version than when the backup was taken.

If data is migrated from one FDB cluster to another, for example if a specialized backup/restore feature is implemented in the application layer, the versionstamp uniqueness and monotonic properties can still be preserved by one of two ways:

> * The new FDB cluster can restore at a higher version than the backup was taken, by setting the ``\xff/minRequiredCommitVersion`` metadata key.

Reviewer: I feel a little bit uncomfortable about advertising this, because the lack of https://github.com/apple/foundationdb/pull/1351 makes it easy to shoot oneself in the foot this way.

Author: Makes sense

> * The 2 byte user version provided by the client can be used to preserve generations across restores.  This method will only provide uniqueness, not sequential serialization, across restores.

Reviewer: You would need to reference the record layer paper to explain generations in this context, and I had to read it a few times to recall that the userbytes go at the end, thus meaning this doesn't give you the ordering.

I think I'd just suggest leaving all of this as "Special care must be taken if moving data between two FoundationDB clusters that contains versionstamps, as the differing cluster versions might break the monotonicity." ?

Author: Oops!  Interesting, I took Alloc's comment at the end of [this discussion](https://forums.foundationdb.org/t/versionstamp-vs-committedversion/600/5) to mean that the user defined bytes were a prefix.  Yes I agree, it's better to leave this vague.
